modules/internal/ChapelDynamicLoading.chpl

```diff
@@ -0,0 +1,797 @@
+    }
+  }
+
+  // A store of of all loaded binaries which lives on LOCALE-0. Each time a
+  // binary is loaded an entry will be made in the store. The interface to the
+  // stored info is the system pointer retrieved by 'dlopen'.
+  class chpl_BinaryInfoStore {
+    forwarding var _lockWrapper = new chpl_lockWrapper();
+
+    // The key is the 'dlopen' pointer returned on LOCALE-0. You are able to
+    // 'dlopen' a symbol multiple times, so you can get the key as needed.
+    var _handleToInfo: chpl_localMap(c_ptr(void), unmanaged chpl_BinaryInfo?);
+
+    inline proc set(ptr: c_ptr(void), bin: unmanaged chpl_BinaryInfo?) {
+      return _handleToInfo.set(ptr, bin);
+    }
+
+    inline proc get(ptr: c_ptr(void), out bin: unmanaged chpl_BinaryInfo?) {
+      return _handleToInfo.get(ptr, bin);
+    }
+
+    proc deinit() {
+      for slot in _handleToInfo.slots() do {
+        if slot.val != nil then delete slot.val;
+        slot.val = nil;
+      }
+    }
+  }
+
+  // This is one per entrypoint binary and lives on LOCALE-0.
+  var chpl_binaryInfoStore = new owned chpl_BinaryInfoStore();
+
+  class DynLibError : Error {
+    proc init(msg: string) {
+      super.init(msg);
+      init this;
+    }
+  }
+
+  // This class represents a "wide" binary. It contains the state necessary
+  // to load a symbol from the binary on each locale.
+  class chpl_BinaryInfo {
+    forwarding var lockWrapper = new chpl_lockWrapper();
+
+    // This refcount is bumped and dropped by the user-facing wrapper.
+    var _refCount: atomic int = 0;
+
+    // This is the path that was used to load the binary.
+    var _path: string;
+
+    // This is the set of loaded binary pointers, indexed by locale. It lives
+    // on LOCALE-0. Since it is a local buffer it can only be accessed there!
+    var _systemPtrs = new chpl_localBuffer(c_ptr(void), numLocales);
+
+    // Local pointers for loaded symbols on LOCALE-0. These are used to evict
+    // entries from the procedure pointer cache when the library is finally
+    // closed. The value is a pair of "wide index" and the symbol name.
+    var _procPtrToDataLocale0: chpl_localMap(c_ptr(void), (int, string));
+
+    // A pointer to the parent store is used to coordinate load/unload.
+    var _store: borrowed chpl_BinaryInfoStore;
+
+    // Set only when '_close()' is called for the first time.
+    var _closed = false;
+
+    proc init(in path: string, in store: borrowed chpl_BinaryInfoStore) {
+      this._path = path;
+      this._store = store;
+      init this;
+    }
+
+    proc deinit() {
+      _close();
+      if _procPtrToDataLocale0.needsDestroy() {
+        import MemMove;
+        for slot in _procPtrToDataLocale0.slots() do MemMove.destroy(slot);
+      }
+    }
+
+    // Load a binary given a path.
+    proc type create(path: string, out err: owned DynLibError?) {
+      var ret: unmanaged chpl_BinaryInfo? = nil;
+
+      on Locales[0] {
+        ref store = chpl_binaryInfoStore;
```

Comment:
isn't this going to create an additional layer of indirection since notionally `chpl_binaryInfoStore` is already a pointer?